File: intrinsic/world/robot_payload/python/robot_payload.py
# Copyright 2023 Intrinsic Innovation LLC

"""The RobotPayload class represents the payload of a kinematic object."""

import math
from intrinsic.math.python import data_types
from intrinsic.math.python import proto_conversion
from intrinsic.world.proto import robot_payload_pb2
import numpy as np

_INERTA_SHAPE = (3, 3)


class RobotPayload:
  """Payload of a kinematic object."""

  def __init__(self):
    # Mass of the robot payload. Unit is kg.
    self._mass: float = 0.0

    # Center of gravity of the robot payload relative to the robot flange/tip
    # frame.
    self._tip_t_cog: data_types.Pose3 = data_types.Pose3()

    # 3x3 symmetric inertia matrix of the robot payload expressed about the link
    # center of mass. Unit is kg*m^2.
    self._inertia: np.ndarray = np.zeros(_INERTA_SHAPE)

  @property
  def mass(self) -> float:
    """Mass of the robot payload. Unit is kg."""
    return self._mass

  @property
  def tip_t_cog(self) -> data_types.Pose3:
    """Center of gravity of the robot payload expressed in the flange frame."""
    return self._tip_t_cog

  @property
  def inertia(self) -> np.ndarray:
    """Inertia of the robot payload expressed about the link center of mass."""
    return self._inertia

  @classmethod
  def create(
      cls,
      mass: float,
      tip_t_cog: data_types.Pose3,
      inertia: np.ndarray,
  ) -> 'RobotPayload':
    """Creates a RobotPayload. Does not check if the payload is valid."""
    payload = RobotPayload()
    payload.set_mass(mass)
    payload.set_tip_t_cog(tip_t_cog)
    payload.set_inertia(inertia)
    return payload

  def set_mass(self, mass: float) -> None:
    """Sets the mass of the robot payload. Unit is kg."""
    self._mass = mass

  def set_tip_t_cog(self, tip_t_cog: data_types.Pose3) -> None:
    """Sets the center of gravity of the robot payload."""
    self._tip_t_cog = tip_t_cog

  def set_inertia(self, inertia: np.ndarray) -> None:
    """Sets the inertia of the robot payload."""
    if inertia.shape != _INERTA_SHAPE:
      raise ValueError(f'Inertia must be a 3x3 matrix, got {inertia.shape}.')
    self._inertia = inertia

  def __eq__(self, other: 'RobotPayload') -> bool:
    if not isinstance(other, RobotPayload):
      return NotImplemented
    return (
        math.isclose(self.mass, other.mass)
        and self.tip_t_cog.almost_equal(other.tip_t_cog)
        and np.allclose(self.inertia, other.inertia)
    )

  def __str__(self):
    return (
        f'RobotPayload(mass={self.mass}, tip_t_cog={self.tip_t_cog},'
        f' inertia={self.inertia})'
    )


def payload_from_proto(
    proto: robot_payload_pb2.RobotPayload,
) -> RobotPayload:
  mass = proto.mass_kg
  tip_t_cog = data_types.Pose3()
  if proto.HasField('tip_t_cog'):
    tip_t_cog = proto_conversion.pose_from_proto(proto.tip_t_cog)
  inertia = np.zeros((3, 3))
  if proto.HasField('inertia'):
    inertia = proto_conversion.ndarray_from_matrix_proto(proto.inertia)

  return RobotPayload.create(
      mass,
      tip_t_cog,
      inertia,
  )


def payload_to_proto(
    payload: RobotPayload,
) -> robot_payload_pb2.RobotPayload:
  return robot_payload_pb2.RobotPayload(
      mass_kg=payload.mass,
      tip_t_cog=proto_conversion.pose_to_proto(payload.tip_t_cog),
      inertia=proto_conversion.ndarray_to_matrix_proto(payload.inertia),
  )

File: intrinsic/world/robot_payload/robot_payload.cc
// Copyright 2023 Intrinsic Innovation LLC

#include "intrinsic/world/robot_payload/robot_payload.h"

#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "intrinsic/eigenmath/types.h"
#include "intrinsic/kinematics/validate_link_parameters.h"
#include "intrinsic/math/pose3.h"
#include "intrinsic/math/proto_conversion.h"
#include "intrinsic/util/status/status_macros.h"
#include "intrinsic/world/proto/robot_payload.pb.h"

namespace intrinsic {

absl::Status RobotPayload::SetMass(double mass_kg) {
  // Allow a zero mass for robots without payload.
  if (!MassAlmostEqual(mass_kg, 0.0)) {
    INTR_RETURN_IF_ERROR(intrinsic::kinematics::ValidateMass(mass_kg));
  }
  mass_kg_ = mass_kg;
  return absl::OkStatus();
}

absl::Status RobotPayload::SetTipTCog(const Pose3d& tip_t_cog) {
  tip_t_cog_ = tip_t_cog;
  return absl::OkStatus();
}

absl::Status RobotPayload::SetInertia(const eigenmath::Matrix3d& inertia) {
  // Allow a zero matrix to handle point masses.
  if (!inertia.isZero()) {
    INTR_RETURN_IF_ERROR(intrinsic::kinematics::ValidateInertia(inertia));
  }

  inertia_in_cog_ = inertia;
  return absl::OkStatus();
}

absl::StatusOr<RobotPayload> RobotPayload::Create(
    double mass, const Pose3d& tip_t_cog, const eigenmath::Matrix3d& inertia) {
  RobotPayload payload;
  INTR_RETURN_IF_ERROR(payload.SetMass(mass));
  INTR_RETURN_IF_ERROR(payload.SetTipTCog(tip_t_cog));
  INTR_RETURN_IF_ERROR(payload.SetInertia(inertia));

  return payload;
}

absl::StatusOr<RobotPayload> FromProto(
    const intrinsic_proto::world::RobotPayload& proto) {
  Pose3d tip_t_cog;
  if (proto.has_tip_t_cog()) {
    INTR_ASSIGN_OR_RETURN(tip_t_cog,
                          intrinsic_proto::FromProto(proto.tip_t_cog()));
  }

  eigenmath::Matrix3d inertia = eigenmath::Matrix3d::Zero();
  if (proto.has_inertia()) {
    if (proto.inertia().rows() != 3 || proto.inertia().cols() != 3) {
      return absl::InvalidArgumentError("Inertia must be 3x3 matrix.");
    }
    INTR_ASSIGN_OR_RETURN(inertia, intrinsic_proto::FromProto(proto.inertia()));
  }

  return RobotPayload::Create(proto.mass_kg(), tip_t_cog, inertia);
}

intrinsic_proto::world::RobotPayload ToProto(const RobotPayload& payload) {
  intrinsic_proto::world::RobotPayload proto;
  proto.set_mass_kg(payload.mass());
  *proto.mutable_tip_t_cog() = ToProto(payload.tip_t_cog());
  *proto.mutable_inertia() = ToProto(payload.inertia());
  return proto;
}

}  // namespace intrinsic
